File: Others/65. Valid Number/65. Valid Number.java
class Solution {
    // [+/- .] e[+/- ] use e divide as two part
    public boolean isNumber(String s) {
        int n = s.length();
        int left = 0, right = n-1;
        while (left < n && s.charAt(left) == ' ') left++;
        while (right >= left && s.charAt(right) == ' ') right--;
        if (left > right) return false;
        String cur = s.substring(left, right+1); // remove the front and rear spcace
        
        // get e count
        int counte = 0;
        int posE = -1;
        for (int i = 0; i < cur.length(); i++) {
            if (cur.charAt(i) == 'e' || cur.charAt(i) == 'E') {
                counte++;
                posE = i;
            }
        }
        if (counte > 1) return false;
        else if (counte == 0) return isOk(cur, 1);
        else return isOk(cur.substring(0, posE), 1) && isOk(cur.substring(posE+1), 0); // left and right of E
    }
    private boolean isOk(String str, int k) {
        // sign
        if (str == null || str.length() == 0) return false;
        int n = str.length();
        for (int i = 0; i < n; i++) {
            char c = str.charAt(i);
            if ( (c == '+' || c == '-') && i != 0 ) return false;
        }
        if (str.charAt(0) == '+' || str.charAt(0) == '-') str = str.substring(1);
        if (str == null || str.length() == 0) return false;
        if (str.length() == 1 && str.charAt(0) =='.') return false;
        int countdot = 0;
        for (int i = 0; i < str.length(); i++) {
            if (str.charAt(i) == '.') countdot++;
            else {
                if (!Character.isDigit(str.charAt(i))) return false;
            }
        }
        return countdot <= k;
    }
}

File: Others/65. Valid Number/65. Valid Number.cpp
class Solution {
public:
    bool isNumber(string s) {
        while (s.size() > 0 && s.back() == ' ') s.pop_back();
        while (s.size() > 0 && s[0] == ' ') s.erase(s.begin());
        if (s == "") return false; // remove the front and rear spcace
        
        // get e count
        int counte = 0;
        int posE = -1;
        for (int i = 0; i < s.size(); i++) {
            if (s[i] == 'e' || s[i] == 'E') {
                counte++;
                posE = i;
            }
        }
        if (counte > 1) return false;
        else if (counte == 0) return isOk(s, 1);
        else return isOk(s.substr(0, posE), 1) && isOk(s.substr(posE+1), 0); // left and right of E
    }
    bool isOk(string str, int k) {
        // sign
        
        int n = str.size();
        for (int i = 0; i < n; i++) {
            if ( (str[i] == '+' || str[i] == '-') && i != 0 ) return false;
        }
        if (str[0] == '+' || str[0] == '-') str.erase(str.begin());
        if (str == "" || str == ".") return false;
        
        int countdot = 0;
        for (int i = 0; i < str.size(); i++) {
            if (str[i] == '.') countdot++;
            else {
                if (!isdigit(str[i])) return false;
            }
        }
        return countdot <= k;
    }
};
